src/maintenance.py

```diff
@@ -0,0 +1,145 @@
+            output = e.response
+            metadata = output.get("metadata", []) or []
+            actions = {}
+            for i, result in enumerate(metadata, 1):
+                actions[f"step-{i}"] = {
+                    "description": result["action"],
+                    "error": result["error"],
+                    "id": result["name"],
+                }
+            logger.error("%s", errors)
+            event.set_results({"actions": actions, "errors": errors, "status": "failure"})
+            event.fail()
+        except Exception as e:
+            logger.error(
+                "Failed to exit maintenance mode for unit '%s': %s", self.charm.unit.name, str(e)
+            )
+            event.set_results({"actions": {}, "errors": str(e), "status": "failure"})
+            event.fail()
+
+    def _enter_maintenance_action(self, event: ops.framework.EventBase) -> None:
+        """Bring the given unit into maintenance mode."""
+        force = event.params.get("force")
+        dry_run = event.params.get("dry-run")
+        set_noout = event.params.get("set-noout")
+        stop_osds = event.params.get("stop-osds")
+        check_only = event.params.get("check-only")
+        ignore_check = event.params.get("ignore-check")
+        if check_only and ignore_check:
+            errors = "check-only and ignore-check cannot be used together"
+            event.set_results({"actions": {}, "errors": errors, "status": "failure"})
+            logger.error(errors)
+            event.fail()
+            return
+
+        try:
+            client = microceph_client.Client.from_socket()
+            output = client.cluster.enter_maintenance_mode(
+                gethostname(), force, dry_run, set_noout, stop_osds, check_only, ignore_check
+            )
+            metadata = output.get("metadata", []) or []
+            actions = {}
+            for i, result in enumerate(metadata, 1):
+                actions[f"step-{i}"] = {
+                    "description": result["action"],
+                    "error": result["error"],
+                    "id": result["name"],
+                }
+            event.set_results({"actions": actions, "errors": "", "status": "success"})
+            metadata = output.get("metadata", []) or []
+            if force:
+                logger.warning(
+                    "Forced to enter maintenance mode for %s, all actions were run but "
+                    "errors were ignored.",
+                    self.charm.unit.name,
+                )
+        except MaintenanceOperationFailedException as e:
+            errors = str(e)
+            output = e.response
+            metadata = output.get("metadata", []) or []
+            actions = {}
+            for i, result in enumerate(metadata, 1):
+                actions[f"step-{i}"] = {
+                    "description": result["action"],
+                    "error": result["error"],
+                    "id": result["name"],
+                }
```

Comment:
This exact same code block has four occurrences in the class. This can be refactored out into a separate method which also logs (DEBUG) the content of the actions dict generated.